src/runner.rs

```diff
@@ -0,0 +1,174 @@
+use std::path::{Path, PathBuf};
+
+use crate::context::Context;
+use crate::error::IxaError;
+use crate::global_properties::ContextGlobalPropertiesExt;
+use crate::random::ContextRandomExt;
+use crate::report::ContextReportExt;
+use clap::{Args, Command, FromArgMatches as _};
+
+#[derive(Args, Debug)]
+pub struct BaseArgs {
+    /// Random seed
+    #[arg(short, long, default_value = "0")]
+    pub seed: u64,
+
+    /// Optional path for a global properties config file
+    #[arg(short, long, default_value = "")]
+    pub config: String,
+
+    /// Optional path for report output
+    #[arg(short, long, default_value = "")]
+    pub output_dir: String,
+}
+
+#[derive(Args)]
+pub struct PlaceholderCustom {}
+
+#[allow(clippy::missing_errors_doc)]
+pub fn run_with_custom_args<A, F>(load: F) -> Result<(), Box<dyn std::error::Error>>
+where
+    A: Args,
+    F: Fn(&mut Context, BaseArgs, Option<A>) -> Result<(), IxaError>,
+{
+    let cli = Command::new("Ixa");
+    let cli = BaseArgs::augment_args(cli);
+    let cli = A::augment_args(cli);
+    let matches = cli.get_matches();
+
+    let base_args_matches = BaseArgs::from_arg_matches(&matches)?;
+    let custom_matches = A::from_arg_matches(&matches)?;
+
+    run_with_args_internal(base_args_matches, Some(custom_matches), load)
+}
+
+#[allow(clippy::missing_errors_doc)]
+pub fn run_with_args<F>(load: F) -> Result<(), Box<dyn std::error::Error>>
+where
+    F: Fn(&mut Context, BaseArgs, Option<PlaceholderCustom>) -> Result<(), IxaError>,
+{
+    let cli = Command::new("Ixa");
+    let cli = BaseArgs::augment_args(cli);
+    let matches = cli.get_matches();
+
+    let base_args_matches = BaseArgs::from_arg_matches(&matches)?;
+
+    run_with_args_internal(base_args_matches, None, load)
+}
+
+fn setup_context(args: &BaseArgs) -> Result<Context, IxaError> {
+    // Instantiate a context
+    let mut context = Context::new();
+
+    // Optionally set global properties from a file
+    if !args.config.is_empty() {
+        println!("Loading global properties from: {}", args.config);
+        let config_path = Path::new(&args.config);
+        context.load_global_properties(config_path)?;
+    }
+
+    // Optionally set output dir for reports
+    if !args.output_dir.is_empty() {
+        let output_dir = PathBuf::from(&args.output_dir);
+        let report_config = context.report_options();
+        report_config.directory(output_dir);
+    }
+
+    context.init_random(args.seed);
+    Ok(context)
+}
+
+fn run_with_args_internal<A, F>(
+    args: BaseArgs,
+    custom_args: Option<A>,
+    load: F,
+) -> Result<(), Box<dyn std::error::Error>>
+where
+    A: Args,
+    F: Fn(&mut Context, BaseArgs, Option<A>) -> Result<(), IxaError>,
+{
+    // Create a context
+    let mut context = setup_context(&args)?;
```

Comment:
Is there a reason not to inline setup_context, which seems to only be called from ehre.